bwt/bwt.go

```diff
@@ -0,0 +1,481 @@
+		skip := bwt.lookupSkipByOffset(fPos)
+		strB.WriteByte(skip.char)
+	}
+
+	return strB.String(), nil
+}
+
+// Len return the length of the sequence used to build the BWT
+func (bwt BWT) Len() int {
+	return bwt.getLenOfOriginalStringWithNullChar() - 1
+}
+
+// GetTransform returns the last column of the BWT transform of the original sequence.
+func (bwt BWT) GetTransform() string {
+	return bwt.lastColumn.reconstruct()
+}
+
+// getFCharPosFromOriginalSequenceCharPos looks up mapping from the original position
+// of the sequence to its corresponding position in the First Column of the BWT
+// NOTE: This clearly isn't ideal. Instead of improving this implementation, this will be replaced with
+// something like r-index in the near future.
+func (bwt BWT) getFCharPosFromOriginalSequenceCharPos(originalPos int) int {
+	for i := range bwt.suffixArray {
+		if bwt.suffixArray[i] == originalPos {
+			return i
+		}
+	}
+	panic("Unable to find the corresponding original position for a character in the original sequence in the suffix array. This should not be possible and indicates a malformed BWT.")
+}
+
+// lfSearch LF Search- Last First Search.
+// Finds the valid range within the BWT index where the provided pattern is possible.
+// If the final range is <= 0, then the pattern does not exist in the original sequence.
+func (bwt BWT) lfSearch(pattern string) interval {
+	searchRange := interval{start: 0, end: bwt.getLenOfOriginalStringWithNullChar()}
+	for i := 0; i < len(pattern); i++ {
+		if searchRange.end-searchRange.start <= 0 {
+			return interval{}
+		}
+
+		c := pattern[len(pattern)-1-i]
+		skip, ok := bwt.lookupSkipByChar(c)
+		if !ok {
+			return interval{}
+		}
+		searchRange.start = skip.openEndedInterval.start + bwt.lastColumn.Rank(c, searchRange.start)
+		searchRange.end = skip.openEndedInterval.start + bwt.lastColumn.Rank(c, searchRange.end)
+	}
+	return searchRange
+}
+
+// lookupSkipByChar looks up a skipEntry by its character in the First Column
+func (bwt BWT) lookupSkipByChar(c byte) (entry skipEntry, ok bool) {
+	for i := range bwt.firstColumnSkipList {
+		if bwt.firstColumnSkipList[i].char == c {
+			return bwt.firstColumnSkipList[i], true
+		}
+	}
+	return skipEntry{}, false
+}
+
+// lookupSkipByOffset looks up a skipEntry based off of an
+// offset of the Fist Column of the BWT.
+func (bwt BWT) lookupSkipByOffset(offset int) skipEntry {
+	if offset > bwt.getLenOfOriginalStringWithNullChar()-1 {
+		msg := fmt.Sprintf("offset [%d] exceeds the max bound of the BWT [%d]", offset, bwt.getLenOfOriginalStringWithNullChar()-1)
+		panic(msg)
+	}
+	if offset < 0 {
+		msg := fmt.Sprintf("offset [%d] exceeds the min bound of the BWT [0]", offset)
+		panic(msg)
+	}
+
+	for i := range bwt.firstColumnSkipList {
+		if bwt.firstColumnSkipList[i].openEndedInterval.start <= offset && offset < bwt.firstColumnSkipList[i].openEndedInterval.end {
+			return bwt.firstColumnSkipList[i]
+		}
+	}
+	panic("figure out what to do here")
```

Comment:
what should be done here?

Reply:
Ty, updated this along with one hiding in the bitvector https://github.com/bebop/poly/pull/411/commits/1cc37557ed1da0cd9b1ef350466d5844ef375f11